include/boost/compute/algorithm/detail/copy_on_device.hpp

```diff
@@ -28,80 +28,92 @@ namespace compute {
 namespace detail {
 
 template<class InputIterator, class OutputIterator>
-class copy_kernel : public meta_kernel
+inline event copy_on_device_cpu(InputIterator first,
+                                OutputIterator result,
+                                size_t count,
+                                command_queue &queue)
 {
-public:
-    copy_kernel(const device &device)
-        : meta_kernel("copy")
-    {
-        m_count = 0;
-
-        typedef typename std::iterator_traits<InputIterator>::value_type input_type;
-
-        boost::shared_ptr<parameter_cache> parameters =
-            detail::parameter_cache::get_global_cache(device);
+    meta_kernel k("copy");
+    const device& device = queue.get_device();
+
+    k <<
+        "uint block = " <<
+            "(uint)ceil(((float)count)/get_global_size(0));\n" <<
```

Comment:
This is the same for every thread. You can pass this as an additional parameter to avoid a divide.